### PR TITLE
abstract duplicated code in ipvs proxier and reduce the comments

### PR DESCRIPTION
**Pull request by Lion-Wei**
**What this PR does / why we need it**:

Abstract duplicated code in ipvs proxier and reduce the comments.

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/proxy/ipvs/proxier.go [hunk 1]
```diff
@@ -1861,22 +1822,3 @@ func openLocalPort(lp *utilproxy.LocalPort) (utilproxy.Closeable, error) {
 	glog.V(2).Infof("Opened local port %s", lp.String())
 	return socket, nil
 }
-
-// ipvs Proxier fall back on iptables when it needs to do SNAT for engress packets
```
Comment (m1093782566) [inline pkg/proxy/ipvs/proxier.go]: Please keep the comment until we have better document.

File: pkg/proxy/ipvs/proxier.go [hunk 2]
```diff
@@ -1735,39 +1729,6 @@ func (proxier *Proxier) cleanLegacyService(atciveServices map[string]bool, curre
 	}
 }
 
-// linkKubeServiceChain will Create chain KUBE-SERVICES and link the chin in PREROUTING and OUTPUT
```
Comment (m1093782566) [inline pkg/proxy/ipvs/proxier.go]: keep the comment?

File: pkg/proxy/ipvs/proxier.go (hunk 1 above)
Comment (Lion-Wei) [reply pkg/proxy/ipvs/proxier.go]: I see, thanks.

File: pkg/proxy/ipvs/proxier.go (hunk 2 above)
Comment (Lion-Wei) [reply pkg/proxy/ipvs/proxier.go]: I abstracted this part function to other place, now added comments to be more specify. : )

File: pkg/proxy/ipvs/proxier.go [hunk 3]
```diff
@@ -780,32 +779,20 @@ func CanUseIPVSProxier(handle KernelHandler, ipsetver IPSetVersioner) (bool, err
...
 // It returns true if an error was encountered. Errors are logged.
 func cleanupIptablesLeftovers(ipt utiliptables.Interface) (encounteredError bool) {
-	// Unlink the services chain.
-	args := []string{
-		"-m", "comment", "--comment", "kubernetes service portals",
-		"-j", string(kubeServicesChain),
-	}
-	for _, tc := range tableChainsWithJumpService {
-		if err := ipt.DeleteRule(tc.table, tc.chain, args...); err != nil {
+	// Unlink the our chains.
```
Comment (m1093782566) [inline pkg/proxy/ipvs/proxier.go]: 'our' is not a good name. 

// Unlink the iptables chains created by IPVS proxier?

File: pkg/proxy/ipvs/proxier.go [hunk 4]
```diff
@@ -996,8 +983,25 @@ func (proxier *Proxier) syncProxyRules() {
...
+			"-j", string(chain.chain),
+		}
+		if _, err := proxier.iptables.EnsureRule(utiliptables.Prepend, chain.table, chain.sourceChain, args...); err != nil {
+			glog.Errorf("Failed to ensure that %s chain %s jumps to %s: %v", chain.table, chain.sourceChain, chain.chain, err)
+			return
+		}
+	}
+
 	// TODO: UT output result
-	// Begin install iptables
```
Comment (m1093782566) [inline pkg/proxy/ipvs/proxier.go]: keep it.

File: pkg/proxy/ipvs/proxier.go [hunk 5]
```diff
@@ -1031,19 +1038,13 @@ func (proxier *Proxier) syncProxyRules() {
...
-		writeLine(proxier.natChains, utiliptables.MakeChainLine(KubeMarkMasqChain))
-	}
 	// Install the kubernetes-specific masquerade mark rule. We use a whole chain for
 	// this so that it is easier to flush and change, for example if the mark
 	// value should ever change.
 	writeLine(proxier.natRules, []string{
 		"-A", string(KubeMarkMasqChain),
 		"-j", "MARK", "--set-xmark", proxier.masqueradeMark,
 	}...)
-	// End install iptables
```
Comment (m1093782566) [inline pkg/proxy/ipvs/proxier.go]: keep it?

File: pkg/proxy/ipvs/proxier.go (hunk 3 above)
Comment (Lion-Wei) [reply pkg/proxy/ipvs/proxier.go]: Done. : )

File: pkg/proxy/ipvs/proxier.go [hunk 6]
```diff
@@ -996,8 +983,25 @@ func (proxier *Proxier) syncProxyRules() {
 
 	glog.V(3).Infof("Syncing ipvs Proxier rules")
 
-	// TODO: UT output result
 	// Begin install iptables
+	// Create chain KUBE-SERVICES and link the chin in PREROUTING and OUTPUT
```
Comment (rramkumar1) [inline pkg/proxy/ipvs/proxier.go]: Nit: "...link the chain..."

**Comment by rramkumar1** [comment]
I'm wondering what the motivation is for this change? It seems like you just moved stuff from one place to another...

**Comment by Lion-Wei** [comment]
@rramkumar1 Thanks for carefully review, fixed.
Actually this pr is to abstract `table/chain/jump service` information into one struct, abstract the create/cleanup of different chains together, and also, reduce some comments. To make it more consistent, more concise, and easier to maintain iptables chains.

File: pkg/proxy/ipvs/proxier.go (hunk 2 above)
Comment (m1093782566) [reply pkg/proxy/ipvs/proxier.go]: We should keep this function?

File: pkg/proxy/ipvs/proxier.go [hunk 7]
```diff
@@ -676,18 +683,19 @@ func (proxier *Proxier) syncProxyRules() {
...
 	// this so that it is easier to flush and change, for example if the mark
 	// value should ever change.
 	writeLine(proxier.natRules, []string{
 		"-A", string(KubeMarkMasqChain),
 		"-j", "MARK", "--set-xmark", proxier.masqueradeMark,
 	}...)
+
+	// Create chain KUBE-FIRE-WALL
+	if err := proxier.createKubeFireWallChain(existingNATChains, proxier.natChains); err != nil {
+		glog.Errorf("Failed to create KUBE-FIRE-WALL chain: %v", err)
```
Comment (thockin) [inline pkg/proxy/ipvs/proxier.go]: FWIW "firewall" is one word in English

File: pkg/proxy/ipvs/proxier.go [hunk 8]
```diff
@@ -80,11 +76,14 @@ const (
 
 // tableChainsWithJumpService is the iptables chains ipvs proxy mode used.
 var tableChainsWithJumpService = []struct {
```
Comment (thockin) [inline pkg/proxy/ipvs/proxier.go]: This name doesn't seem tru any more - it's just a table of jumps.

Needs a better comment, too.

File: pkg/proxy/ipvs/proxier.go [hunk 9]
```diff
@@ -80,11 +76,14 @@ const (
...
-	table utiliptables.Table
-	chain utiliptables.Chain
+	table       utiliptables.Table
```
Comment (thockin) [inline pkg/proxy/ipvs/proxier.go]: I would reorganize this to something like:

```
{
    table       utiliptables.Table
￼    from        utiliptables.Chain
    to            utiliptables.Chain
   comment string
}

**Review by thockin — COMMENTED** [review]
small points to make it more readable

File: pkg/proxy/ipvs/proxier.go (hunk 7 above)
Comment (Lion-Wei) [reply pkg/proxy/ipvs/proxier.go]: You mean `KUBE-FIREWALL` would be a better name?

**Comment by Lion-Wei** [comment]
Since ipvs proxier have changed, I integrated more work in this pr, which include:
1. Assemble all iptables chains and initial rules into one table `iptablesJumpChain`.
2. Use `proxier.ipsetList(map[string]*IPSet)` to replace `proxier.loopbackSet/proxier.clusterIPSet...`, and use of tables of those sets named `ipsetInfo` to initial/destroy/sync all ipsets.
3. Use `ipsetWithIptablesChain` store basic iptables match-ipset-jump information.
4. Fix bug of loadbalancer sourceRange, which may result in firewall chain accidentally drop other traffic.

Please take another review, thanks. @m1093782566 @rramkumar1 @thockin

**Comment by m1093782566** [comment]
@Lion-Wei 

> Fix bug of loadbalancer sourceRange, which may result in firewall chain accidentally drop other traffic.

I would suggest keep this fix in another PR for the sake of easy review.

File: pkg/proxy/ipvs/proxier.go (hunk 7 above)
Comment (rramkumar1) [reply pkg/proxy/ipvs/proxier.go]: @Lion-Wei Yes, "KUBE-FIREWALL"

**Comment by rramkumar1** [comment]
@Lion-Wei Can you rebase?

**Comment by Lion-Wei** [comment]
@rramkumar1 Of cause, sorry for the delay, I was out for vacation. I will rebase this pr ASAP.

**Comment by Lion-Wei** [comment]
@rramkumar1 ,Please take another review, thanks. Abstract work in this pr include: 
1. Use `proxier.ipsetList(map[string]*IPSet)` to replace `proxier.loopbackSet/proxier.clusterIPSet...`, and use of tables of those sets named `ipsetInfo` to initial/destroy/sync all ipsets.
2. Use `iptablesChains` store all iptables chains, use `iptablesJumpChain` store all initial iptables rules. 
3. Abstract all iptables operate to function `writeIptablesRules()` and `createAndLinkeKubeChain()`.

**Pull request merged**